File: healthcare-portal/src/app/core/services/provider-onboarding.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface ProviderOnboarding {
  id: string;
  userId: string;
  firstName: string;
  lastName: string;
  email: string;
  phoneNumber: string;
  specialty: string;
  subSpecialty?: string;
  licenseNumber: string;
  licenseState: string;
  npiNumber?: string;
  deaNumber?: string;
  education?: string;
  workHistory?: string;
  malpracticeInsurance?: string;
  bio?: string;
  profilePhotoUrl?: string;
  governmentIdUrl?: string;
  licenseDocumentUrl?: string;
  certificationDocumentUrl?: string;
  malpracticeInsuranceUrl?: string;
  status: string;
  adminRemarks?: string;
  submittedAt?: string;
  reviewedAt?: string;
  reviewedByUserId?: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateProviderOnboardingDto {
  firstName: string;
  lastName: string;
  email: string;
  phoneNumber: string;
  specialty: string;
  subSpecialty?: string;
  licenseNumber: string;
  licenseState: string;
  npiNumber?: string;
  deaNumber?: string;
  education?: string;
  workHistory?: string;
  malpracticeInsurance?: string;
  bio?: string;
  profilePhotoUrl?: string;
  governmentIdUrl?: string;
  licenseDocumentUrl?: string;
  certificationDocumentUrl?: string;
  malpracticeInsuranceUrl?: string;
}

export interface UpdateProviderOnboardingDto {
  firstName?: string;
  lastName?: string;
  phoneNumber?: string;
  specialty?: string;
  subSpecialty?: string;
  licenseNumber?: string;
  licenseState?: string;
  npiNumber?: string;
  deaNumber?: string;
  education?: string;
  workHistory?: string;
  malpracticeInsurance?: string;
  bio?: string;
  profilePhotoUrl?: string;
  governmentIdUrl?: string;
  licenseDocumentUrl?: string;
  certificationDocumentUrl?: string;
  malpracticeInsuranceUrl?: string;
}

export interface ReviewProviderOnboardingDto {
  status: string;
  adminRemarks?: string;
}

export interface OnboardingStatistics {
  totalOnboardings: number;
  pendingOnboardings: number;
  underReviewOnboardings: number;
  approvedOnboardings: number;
  rejectedOnboardings: number;
  requiresMoreInfoOnboardings: number;
  approvalRate: number;
  averageProcessingTimeDays: number;
}

export interface ApiResponse<T> {
  success: boolean;
  message: string;
  data: T;
  statusCode: number;
  errors?: any;
}

@Injectable({
  providedIn: 'root'
})
export class ProviderOnboardingService {
  private apiUrl = `${environment.apiUrl}/api/provideronboarding`;

  constructor(private http: HttpClient) {}

  // Create new onboarding application
  createOnboarding(createDto: CreateProviderOnboardingDto): Observable<ApiResponse<ProviderOnboarding>> {
    return this.http.post<ApiResponse<ProviderOnboarding>>(`${this.apiUrl}`, createDto);
  }

  // Get onboarding by ID
  getOnboarding(id: string): Observable<ApiResponse<ProviderOnboarding>> {
    return this.http.get<ApiResponse<ProviderOnboarding>>(`${this.apiUrl}/${id}`);
  }

  // Get onboarding by user ID
  getOnboardingByUser(userId: string): Observable<ApiResponse<ProviderOnboarding>> {
    return this.http.get<ApiResponse<ProviderOnboarding>>(`${this.apiUrl}/user/${userId}`);
  }

  // Update onboarding application
  updateOnboarding(id: string, updateDto: UpdateProviderOnboardingDto): Observable<ApiResponse<ProviderOnboarding>> {
    return this.http.put<ApiResponse<ProviderOnboarding>>(`${this.apiUrl}/${id}`, updateDto);
  }

  // Submit onboarding for review
  submitOnboarding(id: string): Observable<ApiResponse<ProviderOnboarding>> {
    return this.http.post<ApiResponse<ProviderOnboarding>>(`${this.apiUrl}/${id}/submit`, {});
  }

  // Review onboarding (Admin only)
  reviewOnboarding(id: string, reviewDto: ReviewProviderOnboardingDto): Observable<ApiResponse<ProviderOnboarding>> {
    return this.http.post<ApiResponse<ProviderOnboarding>>(`${this.apiUrl}/${id}/review`, reviewDto);
  }

  // Get all onboarding applications with optional filtering
  getAllOnboardings(status?: string, page: number = 1, pageSize: number = 50): Observable<ApiResponse<ProviderOnboarding[]>> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('pageSize', pageSize.toString());
    
    if (status) {
      params = params.set('status', status);
    }

    return this.http.get<ApiResponse<ProviderOnboarding[]>>(`${this.apiUrl}`, { params });
  }

  // Get pending onboarding applications
  getPendingOnboardings(): Observable<ApiResponse<ProviderOnboarding[]>> {
    return this.http.get<ApiResponse<ProviderOnboarding[]>>(`${this.apiUrl}/pending`);
  }

  // Get onboarding applications by status
  getOnboardingsByStatus(status: string): Observable<ApiResponse<ProviderOnboarding[]>> {
    return this.http.get<ApiResponse<ProviderOnboarding[]>>(`${this.apiUrl}/status/${status}`);
  }

  // Delete onboarding application
  deleteOnboarding(id: string): Observable<ApiResponse<boolean>> {
    return this.http.delete<ApiResponse<boolean>>(`${this.apiUrl}/${id}`);
  }

  // Get onboarding statistics
  getOnboardingStatistics(): Observable<ApiResponse<OnboardingStatistics>> {
    return this.http.get<ApiResponse<OnboardingStatistics>>(`${this.apiUrl}/statistics`);
  }

  // Helper method to handle API errors
  handleError(error: any): string {
    if (error.error?.message) {
      return error.error.message;
    }
    if (error.message) {
      return error.message;
    }
    return 'An unexpected error occurred';
  }

  // Helper method to check if response is successful
  isSuccessResponse(response: ApiResponse<any>): boolean {
    return response.success && response.statusCode >= 200 && response.statusCode < 300;
  }
} 

File: backend/SmartTelehealth.Application/Services/ProviderOnboardingService.cs
using AutoMapper;
using SmartTelehealth.Application.DTOs;
using SmartTelehealth.Application.Interfaces;
using SmartTelehealth.Core.Entities;
using SmartTelehealth.Core.Interfaces;
using SmartTelehealth.Application.Services;
using System.Security.Claims;
using Microsoft.Extensions.Logging;

namespace SmartTelehealth.Application.Services;

public class ProviderOnboardingService : IProviderOnboardingService
{
    private readonly IProviderOnboardingRepository _onboardingRepository;
    private readonly IUserService _userService;
    private readonly IAuditService _auditService;
    private readonly INotificationService _notificationService;
    private readonly IMapper _mapper;
    private readonly ILogger<ProviderOnboardingService> _logger;

    public ProviderOnboardingService(
        IProviderOnboardingRepository onboardingRepository,
        IUserService userService,
        IAuditService auditService,
        INotificationService notificationService,
        IMapper mapper,
        ILogger<ProviderOnboardingService> logger)
    {
        _onboardingRepository = onboardingRepository;
        _userService = userService;
        _auditService = auditService;
        _notificationService = notificationService;
        _mapper = mapper;
        _logger = logger;
    }

    public async Task<JsonModel> CreateOnboardingAsync(CreateProviderOnboardingDto createDto)
    {
        try
        {
            // TODO: Check if user already has an onboarding application by email
            // This will be implemented when we have proper user management

            var onboarding = new ProviderOnboarding
            {
                Id = Guid.NewGuid(),
                UserId = 0, // TODO: This should be set to the actual user ID after user creation
                FirstName = createDto.FirstName,
                LastName = createDto.LastName,
                Email = createDto.Email,
                PhoneNumber = createDto.PhoneNumber,
                Specialty = createDto.Specialty,
                SubSpecialty = createDto.SubSpecialty,
                LicenseNumber = createDto.LicenseNumber,
                LicenseState = createDto.LicenseState,
                NPINumber = createDto.NPINumber,
                DEANumber = createDto.DEANumber,
                Education = createDto.Education,
                WorkHistory = createDto.WorkHistory,
                MalpracticeInsurance = createDto.MalpracticeInsurance,
                Bio = createDto.Bio,
                ProfilePhotoUrl = createDto.ProfilePhotoUrl,
                GovernmentIdUrl = createDto.GovernmentIdUrl,
                LicenseDocumentUrl = createDto.LicenseDocumentUrl,
                CertificationDocumentUrl = createDto.CertificationDocumentUrl,
                MalpracticeInsuranceUrl = createDto.MalpracticeInsuranceUrl,
                Status = OnboardingStatus.Pending,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };

            var savedOnboarding = await _onboardingRepository.AddAsync(onboarding);
            var dto = _mapper.Map<ProviderOnboardingDto>(savedOnboarding);

            // TODO: Implement audit logging
            // await _auditService.LogActionAsync("ProviderOnboarding", "Create", savedOnboarding.Id.ToString(), "Onboarding application created");

            return new JsonModel
            {
                data = dto,
                Message = "Onboarding application created successfully",
                StatusCode = 201
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error creating onboarding application");
            return new JsonModel
            {
                data = new object(),
                Message = "Error creating onboarding application",
                StatusCode = 500
            };
        }
    }

    public async Task<JsonModel> GetOnboardingAsync(Guid id)
    {
        try
        {
            var onboarding = await _onboardingRepository.GetByIdAsync(id);
            if (onboarding == null)
            {
                return new JsonModel
                {
                    data = new object(),
                    Message = "Onboarding application not found",
                    StatusCode = 404
                };
            }

            var dto = _mapper.Map<ProviderOnboardingDto>(onboarding);
            return new JsonModel
            {
                data = dto,
                Message = "Onboarding application retrieved successfully",
                StatusCode = 200
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving onboarding application");
            return new JsonModel
            {
                data = new object(),
                Message = "Error retrieving onboarding application",
                StatusCode = 500
            };
        }
    }

    public async Task<JsonModel> GetOnboardingByUserIdAsync(int userId)
    {
        try
        {
            var onboarding = await _onboardingRepository.GetByUserIdAsync(userId);
            if (onboarding == null)
            {
                return new JsonModel
                {
                    data = new object(),
                    Message = "Onboarding application not found",
                    StatusCode = 404
                };
            }

            var dto = _mapper.Map<ProviderOnboardingDto>(onboarding);
            return new JsonModel
            {
                data = dto,
                Message = "Onboarding application retrieved successfully",
                StatusCode = 200
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving onboarding application by user ID");
            return new JsonModel
            {
                data = new object(),
                Message = "Error retrieving onboarding application",
                StatusCode = 500
            };
        }
    }

    public async Task<JsonModel> UpdateOnboardingAsync(Guid id, UpdateProviderOnboardingDto updateDto)
    {
        try
        {
            var onboarding = await _onboardingRepository.GetByIdAsync(id);
            if (onboarding == null)
            {
                return new JsonModel
                {
                    data = new object(),
                    Message = "Onboarding application not found",
                    StatusCode = 404
                };
            }

            // Update fields if provided
            if (!string.IsNullOrEmpty(updateDto.FirstName))
                onboarding.FirstName = updateDto.FirstName;
            if (!string.IsNullOrEmpty(updateDto.LastName))
                onboarding.LastName = updateDto.LastName;
            if (!string.IsNullOrEmpty(updateDto.PhoneNumber))
                onboarding.PhoneNumber = updateDto.PhoneNumber;
            if (!string.IsNullOrEmpty(updateDto.Specialty))
                onboarding.Specialty = updateDto.Specialty;
            if (!string.IsNullOrEmpty(updateDto.SubSpecialty))
                onboarding.SubSpecialty = updateDto.SubSpecialty;
            if (!string.IsNullOrEmpty(updateDto.LicenseNumber))
                onboarding.LicenseNumber = updateDto.LicenseNumber;
            if (!string.IsNullOrEmpty(updateDto.LicenseState))
                onboarding.LicenseState = updateDto.LicenseState;
            if (!string.IsNullOrEmpty(updateDto.NPINumber))
                onboarding.NPINumber = updateDto.NPINumber;
            if (!string.IsNullOrEmpty(updateDto.DEANumber))
                onboarding.DEANumber = updateDto.DEANumber;
            if (!string.IsNullOrEmpty(updateDto.Education))
                onboarding.Education = updateDto.Education;
            if (!string.IsNullOrEmpty(updateDto.WorkHistory))
                onboarding.WorkHistory = updateDto.WorkHistory;
            if (!string.IsNullOrEmpty(updateDto.MalpracticeInsurance))
                onboarding.MalpracticeInsurance = updateDto.MalpracticeInsurance;
            if (!string.IsNullOrEmpty(updateDto.Bio))
                onboarding.Bio = updateDto.Bio;
            if (!string.IsNullOrEmpty(updateDto.ProfilePhotoUrl))
                onboarding.ProfilePhotoUrl = updateDto.ProfilePhotoUrl;
            if (!string.IsNullOrEmpty(updateDto.GovernmentIdUrl))
                onboarding.GovernmentIdUrl = updateDto.GovernmentIdUrl;
            if (!string.IsNullOrEmpty(updateDto.LicenseDocumentUrl))
                onboarding.LicenseDocumentUrl = updateDto.LicenseDocumentUrl;
            if (!string.IsNullOrEmpty(updateDto.CertificationDocumentUrl))
                onboarding.CertificationDocumentUrl = updateDto.CertificationDocumentUrl;
            if (!string.IsNullOrEmpty(updateDto.MalpracticeInsuranceUrl))
                onboarding.MalpracticeInsuranceUrl = updateDto.MalpracticeInsuranceUrl;

            onboarding.UpdatedAt = DateTime.UtcNow;

            var updatedOnboarding = await _onboardingRepository.UpdateAsync(onboarding);
            var dto = _mapper.Map<ProviderOnboardingDto>(updatedOnboarding);

            // TODO: Implement audit logging
            // await _auditService.LogActionAsync("ProviderOnboarding", "Update", id.ToString(), "Onboarding application updated");

            return new JsonModel
            {
                data = dto,
                Message = "Onboarding application updated successfully",
                StatusCode = 200
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error updating onboarding application");
            return new JsonModel
            {
                data = new object(),
                Message = "Error updating onboarding application",
                StatusCode = 500
            };
        }
    }

    public async Task<JsonModel> SubmitOnboardingAsync(Guid id)
    {
        try
        {
            var onboarding = await _onboardingRepository.GetByIdAsync(id);
            if (onboarding == null)
            {
                return new JsonModel
                {
                    data = new object(),
                    Message = "Onboarding application not found",
                    StatusCode = 404
                };
            }

            if (onboarding.Status != OnboardingStatus.Pending)
            {
                return new JsonModel
                {
                    data = new object(),
                    Message = "Onboarding application cannot be submitted in current status",
                    StatusCode = 400
                };
            }

            onboarding.Status = OnboardingStatus.UnderReview;
            onboarding.SubmittedAt = DateTime.UtcNow;
            onboarding.UpdatedAt = DateTime.UtcNow;

            var updatedOnboarding = await _onboardingRepository.UpdateAsync(onboarding);
            var dto = _mapper.Map<ProviderOnboardingDto>(updatedOnboarding);

            // Log audit
            // TODO: Implement audit logging
            // await _auditService.LogActionAsync("ProviderOnboarding", "Submit", id.ToString(), "Onboarding application submitted for review");

            // TODO: Implement notification service
            // await _notificationService.SendNotificationAsync(
            //     "New Onboarding Application",
            //     $"New provider onboarding application submitted by {onboarding.FirstName} {onboarding.LastName}",
            //     "Admin"
            // );

            return new JsonModel
            {
                data = dto,
                Message = "Onboarding application submitted successfully",
                StatusCode = 200
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error submitting onboarding application");
            return new JsonModel
            {
                data = new object(),
                Message = "Error submitting onboarding application",
                StatusCode = 500
            };
        }
    }

    public async Task<JsonModel> ReviewOnboardingAsync(Guid id, ReviewProviderOnboardingDto reviewDto)
    {
        try
        {
            var onboarding = await _onboardingRepository.GetByIdAsync(id);
            if (onboarding == null)
            {
                return new JsonModel
                {
                    data = new object(),
                    Message = "Onboarding application not found",
                    StatusCode = 404
                };
            }

            if (onboarding.Status != OnboardingStatus.UnderReview && onboarding.Status != OnboardingStatus.RequiresMoreInfo)
            {
                return new JsonModel
                {
                    data = new object(),
                    Message = "Onboarding application cannot be reviewed in current status",
                    StatusCode = 400
                };
            }

            // Update status based on review
            onboarding.Status = Enum.Parse<OnboardingStatus>(reviewDto.Status);
            onboarding.AdminRemarks = reviewDto.AdminRemarks;
            onboarding.ReviewedAt = DateTime.UtcNow;
            onboarding.UpdatedAt = DateTime.UtcNow;

            var updatedOnboarding = await _onboardingRepository.UpdateAsync(onboarding);
            var dto = _mapper.Map<ProviderOnboardingDto>(updatedOnboarding);

            // Log audit
            // TODO: Implement audit logging
            // await _auditService.LogActionAsync("ProviderOnboarding", "Review", id.ToString(), $"Onboarding application reviewed: {reviewDto.Status}");

            // Send notification to provider
            var notificationMessage = reviewDto.Status == "Approved" 
                ? "Your onboarding application has been approved! You can now access the provider portal."
                : $"Your onboarding application has been {reviewDto.Status.ToLower()}. {reviewDto.AdminRemarks}";

            // TODO: Implement notification service
            // await _notificationService.SendNotificationAsync(
            //     "Onboarding Application Update",
            //     notificationMessage,
            //     onboarding.UserId.ToString()
            // );

            return new JsonModel
            {
                data = dto,
                Message = "Onboarding application reviewed successfully",
                StatusCode = 200
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error reviewing onboarding application");
            return new JsonModel
            {
                data = new object(),
                Message = "Error reviewing onboarding application",
                StatusCode = 500
            };
        }
    }

    public async Task<JsonModel> GetAllOnboardingsAsync(string? status = null, int page = 1, int pageSize = 50)
    {
        try
        {
            IEnumerable<ProviderOnboarding> onboardings;
            
            if (!string.IsNullOrEmpty(status))
            {
                onboardings = await _onboardingRepository.GetByStatusWithPaginationAsync(status, page, pageSize);
            }
            else
            {
                onboardings = await _onboardingRepository.GetAllAsync();
            }

            var dtos = _mapper.Map<IEnumerable<ProviderOnboardingDto>>(onboardings);
            return new JsonModel
            {
                data = dtos,
                Message = "Onboarding applications retrieved successfully",
                StatusCode = 200
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving onboarding applications");
            return new JsonModel
            {
                data = new object(),
                Message = "Error retrieving onboarding applications",
                StatusCode = 500
            };
        }
    }

    public async Task<JsonModel> GetPendingOnboardingsAsync()
    {
        try
        {
            var onboardings = await _onboardingRepository.GetPendingAsync();
            var dtos = _mapper.Map<IEnumerable<ProviderOnboardingDto>>(onboardings);
            return new JsonModel
            {
                data = dtos,
                Message = "Pending onboarding applications retrieved successfully",
                StatusCode = 200
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving pending onboarding applications");
            return new JsonModel
            {
                data = new object(),
                Message = "Error retrieving pending onboarding applications",
                StatusCode = 500
            };
        }
    }

    public async Task<JsonModel> GetOnboardingsByStatusAsync(string status)
    {
        try
        {
            var onboardings = await _onboardingRepository.GetByStatusAsync(status);
            var dtos = _mapper.Map<IEnumerable<ProviderOnboardingDto>>(onboardings);
            return new JsonModel
            {
                data = dtos,
                Message = "Onboarding applications by status retrieved successfully",
                StatusCode = 200
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving onboarding applications by status");
            return new JsonModel
            {
                data = new object(),
                Message = "Error retrieving onboarding applications by status",
                StatusCode = 500
            };
        }
    }

    public async Task<JsonModel> DeleteOnboardingAsync(Guid id)
    {
        try
        {
            var result = await _onboardingRepository.DeleteAsync(id);
            if (!result)
            {
                return new JsonModel
                {
                    data = new object(),
                    Message = "Onboarding application not found",
                    StatusCode = 404
                };
            }

            // Log audit
            // TODO: Implement audit logging
            // await _auditService.LogActionAsync("ProviderOnboarding", "Delete", id.ToString(), "Onboarding application deleted");

            return new JsonModel
            {
                data = true,
                Message = "Onboarding application deleted successfully",
                StatusCode = 200
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error deleting onboarding application");
            return new JsonModel
            {
                data = new object(),
                Message = "Error deleting onboarding application",
                StatusCode = 500
            };
        }
    }

    public async Task<JsonModel> GetOnboardingStatisticsAsync()
    {
        try
        {
            var totalOnboardings = await _onboardingRepository.GetTotalCountAsync();
            var pendingOnboardings = await _onboardingRepository.GetCountByStatusAsync("Pending");
            var underReviewOnboardings = await _onboardingRepository.GetCountByStatusAsync("UnderReview");
            var approvedOnboardings = await _onboardingRepository.GetCountByStatusAsync("Approved");
            var rejectedOnboardings = await _onboardingRepository.GetCountByStatusAsync("Rejected");
            var requiresMoreInfoOnboardings = await _onboardingRepository.GetCountByStatusAsync("RequiresMoreInfo");

            var approvalRate = totalOnboardings > 0 
                ? (decimal)(approvedOnboardings + rejectedOnboardings) > 0 
                    ? (decimal)approvedOnboardings / (decimal)(approvedOnboardings + rejectedOnboardings) * 100 
                    : 0 
                : 0;

            var statistics = new OnboardingStatisticsDto
            {
                TotalOnboardings = totalOnboardings,
                PendingOnboardings = pendingOnboardings,
                UnderReviewOnboardings = underReviewOnboardings,
                ApprovedOnboardings = approvedOnboardings,
                RejectedOnboardings = rejectedOnboardings,
                RequiresMoreInfoOnboardings = requiresMoreInfoOnboardings,
                ApprovalRate = approvalRate,
                AverageProcessingTimeDays = 3 // TODO: Calculate actual average
            };

            return new JsonModel
            {
                data = statistics,
                Message = "Onboarding statistics retrieved successfully",
                StatusCode = 200
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving onboarding statistics");
            return new JsonModel
            {
                data = new object(),
                Message = "Error retrieving onboarding statistics",
                StatusCode = 500
            };
        }
    }
} 